File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def maxSumBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        self.max_sum=0

        def dfs(ptr):
            if not ptr:
                return (True,float('inf'),float('-inf'),0)
            L_isBST,L_min,L_max,L_sum=dfs(ptr.left)
            R_isBST,R_min,R_max,R_sum=dfs(ptr.right)
            if L_isBST and R_isBST and L_max<ptr.val and ptr.val<R_min:
                curr_sum=R_sum+L_sum+ptr.val
                self.max_sum=max(self.max_sum,curr_sum)
                return (True,min(L_min,ptr.val),max(R_max,ptr.val),curr_sum)
            else:
                return (False,0,0,0)

        dfs(root)
        return self.max_sum



        

File: 1373-maximum-sum-bst-in-binary-tree/1373-maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int max_sum=INT_MIN;
    vector<int> helper(TreeNode* node){
        if(!node){
            return {1,INT_MAX,INT_MIN,0};
        }
        auto l=helper(node->left);
        int l_bst=l[0];
        int l_min=l[1];
        int l_max=l[2];
        int l_sum=l[3];

        auto r=helper(node->right);
        int r_bst=r[0];
        int r_min=r[1];
        int r_max=r[2];
        int r_sum=r[3];

        if(l_bst && r_bst && l_max<node->val && r_min>node->val){
            int curr_sum=l_sum+r_sum+node->val;
            max_sum=max(max_sum,curr_sum);
            return {1,min(l_min,node->val), max(r_max,node->val),curr_sum};
        }
        else{
            return {0,0,0,0};
        }
    }
    int maxSumBST(TreeNode* root) {
        helper(root);
        return max(0,max_sum);
    }
};